### PR TITLE
Custom properties feature

### PR DESCRIPTION
**Pull request by MatkovIvan**
Implemented custom properties feature by spec

### CODE REVIEW

**Comment by jaelim-ms** [comment]
Something goes wrong on test. Can you run unit test locally and see if there are any crashes even though unit test succeeds? This is one possibility that messes up coverage.

File: MobileCenter/MobileCenter/Internals/MSCustomPropertiesPrivate.h [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/**
+ *  Protected declarations for MSCustomProperties.
+ */
+@interface MSCustomProperties ()
+
+@property(nonatomic, retain) NSMutableDictionary<NSString *, NSObject *> *properties;
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/MSCustomPropertiesPrivate.h]: please use `strong` instead of `retain`

File: MobileCenter/MobileCenter/Internals/Model/MSCustomPropertiesLog.m [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        [property setObject:key forKey:kMSPropertyName];
+        [propertiesArray addObject:property];
+      }
+    }
+    dict[kMSProperties] = propertiesArray;
+  }  return dict;
+}
+
++ (NSMutableDictionary *)serializeProperty:(NSObject *)value {
+  static NSDateFormatter *dateFormatter = nil;
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Model/MSCustomPropertiesLog.m]: I normally don't like this but imho in this case, it makes sense to have the dateFormatter `static`.

File: MobileCenter/MobileCenter/MSCustomProperties.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
+#import <Foundation/Foundation.h>
+
+/**
+ * Custom properties builder.
+ * Collect multiple properties for send its at once in the same log.
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/MSCustomProperties.h]: image

File: MobileCenter/MobileCenter/MSMobileCenter.h [hunk 4]
```diff
@@ -2,6 +2,7 @@
 #import "MSConstants.h"
 
 @class MSWrapperSdk;
+@class MSCustomProperties;
 
 /**
  * Class comment: Some Introduction.
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/MSMobileCenter.h]: Cleanup or add a propper comment.

**Comment by TroubleMakerBen** [comment]
@MatkovIvan nice, please address https://github.com/Microsoft/mobile-center-sdk-ios/pull/449/files/69589a248af26e0ae34037fa55c68cf0556c00fa#diff-43ee3846270cc1f5492d0f308bf6e7e5R8 and we're good to go.

File: MobileCenter/MobileCenter/MSMobileCenter.h (hunk 4 above)
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/MSMobileCenter.h]: This comment shouldn't be here.

Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/MSMobileCenter.h]: Not part of this PR.

File: MobileCenter/MobileCenter/MSCustomProperties.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Custom properties builder.
+ * Collects multiple properties to send in one log.
+ */
+@interface MSCustomProperties : NSObject
+
+/**
+ * Set the specified property value with the specified key.
+ * If the properties previously contained a property for the key, the old
+ * value is replaced.
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSCustomProperties.h]: The above line is not long. make a single line for the sentence.

File: MobileCenter/MobileCenter/MSCustomProperties.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * @param key   key with which the specified value is to be set.
+ * @param value value to be set with the specified key.
+ * @return this instance.
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSCustomProperties.h]: Space before return

File: MobileCenter/MobileCenter/MSCustomProperties.h (hunk 5 above)
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSCustomProperties.h]: The other methods, too.

File: MobileCenter/MobileCenter/MSCustomProperties.h (hunk 6 above)
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSCustomProperties.h]: The other methods, too.

File: MobileCenter/MobileCenter/MSCustomProperties.m [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+- (MSCustomProperties *)setString:(NSString *)value forKey:(NSString *)key {
+  if ([self isValidKey:key]) {
+    if (value) {
+      [self.properties setObject:value forKey:key];
+    } else {
+      MSLogError([MSMobileCenter logTag], kValueNullErrorMessage);
+    }
+  }
+  return self;
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSCustomProperties.m]: Add a log that key is not valid.

Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSCustomProperties.m]: The others, too.

File: MobileCenter/MobileCenter/MSCustomProperties.m [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+- (instancetype)init {
+  if ((self = [super init])) {
+    _properties = [NSMutableDictionary new];
+  }
+  return self;
+}
+
+- (MSCustomProperties *)setString:(NSString *)value forKey:(NSString *)key {
+  if ([self isValidKey:key]) {
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSCustomProperties.m]: This is statement is repeating. Separate method?

File: MobileCenter/MobileCenter/MSCustomProperties.m [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    [self.properties setObject:[NSNull null] forKey:key];
+  }
+  return self;
+}
+
+- (BOOL) isValidKey:(NSString *)key {
+  static NSRegularExpression *regex = nil;
+  if (!regex) {
+    NSError *error = nil;
+    regex = [NSRegularExpression regularExpressionWithPattern:kKeyPattern options:0 error:&error];
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSCustomProperties.m]: No compiler warning for the value of `options`?

File: MobileCenter/MobileCenter/MSMobileCenter.m [hunk 10]
```diff
@@ -249,6 +256,19 @@ - (void)setLogUrl:(NSString *)logUrl {
   }
 }
 
+- (void)setCustomProperties:(MSCustomProperties *)customProperties {
+  if (!customProperties) {
+    MSLogError([MSMobileCenter logTag], @"Custom properties may not be null");
+    return;
+  }
+  NSDictionary<NSString *, NSObject *> *properties = customProperties.properties;
+  if (properties.count == 0) {
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSMobileCenter.m]: Combine with above if statement. You can just say *Custom properties may not be null or empty*

File: MobileCenter/MobileCenter/MSMobileCenter.m [hunk 11]
```diff
@@ -353,6 +373,12 @@ - (void)sendStartServiceLog:(NSArray<NSString *> *)servicesNames {
   [self.logManager processLog:serviceLog withPriority:MSPriorityDefault andGroupID:kMSGroupID];
 }
 
+- (void)sendCustomPropertiesLog:(NSDictionary<NSString *, NSObject *> *)properties {
+  MSCustomPropertiesLog *customPropertiesLog = [MSCustomPropertiesLog new];
+  customPropertiesLog.properties = properties;
+  [self.logManager processLog:customPropertiesLog withPriority:MSPriorityDefault andGroupID:kMSGroupID];
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSMobileCenter.m]: This will be a merge conflict with #444. `withProiority:MSPriorityDefault` need to be removed if #444 goes first.
Good to have TODO or FIXME

File: MobileCenter/MobileCenter/Internals/Model/MSCustomPropertiesLog.h [hunk 12]
```diff
@@ -0,0 +1,11 @@
+#import "MSAbstractLog.h"
+#import <Foundation/Foundation.h>
+
+@interface MSCustomPropertiesLog : MSAbstractLog
+
+/**
+ * Properties key/value pairs.
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Model/MSCustomPropertiesLog.h]: *Key/value pair properties*?

File: MobileCenter/MobileCenter/Internals/Model/MSCustomPropertiesLog.m [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    [dateFormatter setLocale:[NSLocale systemLocale]];
+    [dateFormatter setTimeZone:[NSTimeZone timeZoneWithAbbreviation: @"UTC"]];
+    [dateFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss'Z'"];
+  }
+  NSMutableDictionary *property = [NSMutableDictionary new];
+  if ([value isKindOfClass:[NSNull class]]) {
+    [property setObject:kMSPropertyTypeClear forKey:kMSPropertyType];
+  } else if ([value isKindOfClass:[NSNumber class]]) {
+    NSNumber * numberValue = (NSNumber *)value;
+    if (!strcmp([numberValue objCType], @encode(BOOL))) {
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Model/MSCustomPropertiesLog.m]: Not sure what you expected from `objCType` for BOOL here. 'c' or 'B'?

**Comment by TroubleMakerBen** [comment]
Don't merge just yet, wait because of #reasons ;)

File: MobileCenter/MobileCenterTests/MSCustomPropertiesTests.m [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+  assertThat([customProperties properties], hasCountOf(0));
+  
+  // Normal value.
+  // When
+  NSString *normalValue = @"test";
+  [customProperties setString:normalValue forKey:key];
+  
+  // Then
+  assertThat([customProperties properties], hasCountOf(1));
+}
```
Comment (clpolet) [inline MobileCenter/MobileCenterTests/MSCustomPropertiesTests.m]: There is the special case of the empty string which is allowed as a value that you may want to add.

File: MobileCenter/MobileCenterTests/MSCustomPropertiesTests.m [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+  [customProperties setDate:date forKey:@"t2"];
+  [customProperties setNumber:number forKey:@"t3"];
+  [customProperties setBool:boolean forKey:@"t4"];
+  [customProperties clearPropertyForKey:@"t5"];
+  
+  // Then
+  assertThat([customProperties properties], hasCountOf(5));
+}
+
+- (void)testSetString {
```
Comment (clpolet) [inline MobileCenter/MobileCenterTests/MSCustomPropertiesTests.m]: While testing the setters, it may be nice to check the value within the properties so that it's what is expected.

File: MobileCenter/MobileCenter/MSCustomProperties.m (hunk 8 above)
Comment (MatkovIvan) [inline MobileCenter/MobileCenter/MSCustomProperties.m]: Repeating code in `isValidKey` method.

File: MobileCenter/MobileCenter/MSCustomProperties.m (hunk 7 above)
Comment (MatkovIvan) [inline MobileCenter/MobileCenter/MSCustomProperties.m]: Log in `isValidKey` method.

File: MobileCenter/MobileCenter/MSMobileCenter.m (hunk 11 above)
Comment (MatkovIvan) [inline MobileCenter/MobileCenter/MSMobileCenter.m]: This PR target to push branch and unlike #444. I added FIXME comment.

Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSMobileCenter.m]: Good enough

File: MobileCenter/MobileCenter/MSCustomProperties.m (hunk 7 above)
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSCustomProperties.m]: OK.

File: MobileCenter/MobileCenter/MSCustomProperties.m (hunk 8 above)
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSCustomProperties.m]: I meant you can introduce private setObject and then use this for setString, setNumber and setDate.

File: MobileCenter/MobileCenter/MSCustomProperties.m [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    if (value) {
+      [self.properties setObject:value forKey:key];
+    } else {
+      MSLogError([MSMobileCenter logTag], kValueNullErrorMessage);
+    }
+  }
+  return self;
+}
+
+- (MSCustomProperties *)setNumber:(NSNumber *)value forKey:(NSString *)key {
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/MSCustomProperties.m]: I meant you can introduce private setObject and then use this for setString, setNumber and setDate.

Comment (MatkovIvan) [inline MobileCenter/MobileCenter/MSCustomProperties.m]: @jaelim-ms Done.
